qoolqit/pulserbackend.py

```diff
@@ -0,0 +1,123 @@
+    and Pasqal's backends (including QPUs and local/remote emulators).
+
+    Args:
+        backend_type (type): backend type from PulserBackend.available_backends().
+        emulator_config (EmulationConfig): optional configuration object to configure
+            remote and local emulators.
+        connection (RemoteConnection): connection to execute the program on remote backends.
+        mimic_qpu (bool):  Whether to mimic the validations necessary for
+            execution on a QPU.
+
+    Examples:
+        TODO:
+         - local and remote example
+         - links to documentation
+    """
+
+    # default emulator configuration to return final time bitstrings
+    default_config = EmulationConfig(observables=(BitStrings(),), log_level=logging.WARN)
+
+    def __init__(
+        self,
+        *,
+        backend_type: type[Backend] = QutipBackendV2,
+        emulation_config: EmulationConfig | None = None,
+        connection: RemoteConnection | None = None,
+        mimic_qpu: bool = False,
+    ) -> None:
+
+        self.backend_type: type[Backend] = self.validate_backend_type(backend_type)
+        if issubclass(self.backend_type, RemoteBackend):
+            self.connection: RemoteConnection = self.validate_connection(connection)
+        self.emulation_config = emulation_config if emulation_config else self.default_config
+        self.mimic_qpu: bool = mimic_qpu
+
+    @staticmethod
+    def validate_backend_type(backend_type: type[Backend]) -> type[Backend]:
+        """Only suppport Pulser EmulatorBackend or RemoteBackend subclasses backend type."""
+        if not issubclass(backend_type, (EmulatorBackend, RemoteBackend)):
+            raise TypeError(f"{backend_type.__name__} is not a supported backend type.")
+        return backend_type
+
+    def validate_connection(self, connection: RemoteConnection | None) -> RemoteConnection:
+        """Validate the required PasqalCloud connection for remote backends.
+
+        Note:
+            Local emulators and internal QPUs require a `PasqalCloud` connection to send jobs.
+            Client's QPUs require a derived OVHConnection(PasqalCloud).
+        """
+        if not isinstance(connection, RemoteConnection):
+            raise TypeError(
+                f"""Error in `PulserBackend`: a remote backend of type
+                {self.backend_type.__name__} requires a `connection` of type {RemoteConnection}."""
+            )
+        if issubclass(self.backend_type, RemoteEmulatorBackend):
+            if not isinstance(connection, PasqalCloud):
+                raise TypeError(
+                    f"""Error in `PulserBackend`: a remote backend of type
+                    {self.backend_type.__name__} requires a `connection` of type {PasqalCloud}."""
+                )
+
+        return connection
+
+    def run(self, sequence: PulserSequence) -> Results | Sequence[Results]:
```

Comment:
Since a `Results` is essentially a `Sequence[Results]` of length 1, maybe simply

```suggestion
    def run(self, sequence: PulserSequence) -> Sequence[Results]:
```

Reply:
`-> Results | Sequence[Results]` is just the signature of the run method in local emulators.